File: sdks/java/sqlflow-postgres/src/main/java/de/bytefish/sqlflow/postgres/PostgresFlowDatabase.java
package de.bytefish.sqlflow.postgres;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import de.bytefish.sqlflow.core.db.SqlFlowDatabase;
import de.bytefish.sqlflow.core.exceptions.CancelledTaskException;
import de.bytefish.sqlflow.core.models.CheckpointRow;
import de.bytefish.sqlflow.core.models.ClaimedTask;
import de.bytefish.sqlflow.core.models.SpawnResult;
import org.postgresql.util.PSQLException;

import java.sql.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * PostgreSQL implementation. Completely synchronous, meant to be consumed by Java 21 Virtual Threads.
 * CancellationToken semantics are inherently mapped to standard Java Thread interruption.
 */
public class PostgresFlowDatabase implements SqlFlowDatabase {

    private final ObjectMapper mapper = new ObjectMapper();

    @Override
    public void createQueue(Connection conn, String queueName) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.create_queue(?, 'unpartitioned')")) {
                cmd.setString(1, queueName);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public void dropQueue(Connection conn, String queueName) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.drop_queue(?)")) {
                cmd.setString(1, queueName);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public List<String> listQueues(Connection conn) {
        return executeWithCancelCheck(() -> {
            List<String> results = new ArrayList<>();
            try (PreparedStatement cmd = conn.prepareStatement("SELECT queue_name FROM ssf.queues ORDER BY queue_name");
                 ResultSet rs = cmd.executeQuery()) {
                while (rs.next()) {
                    results.add(rs.getString(1));
                }
                return results;
            }
        });
    }

    @Override
    public SpawnResult spawnTask(Connection conn, String queue, String taskName, String paramsJson, String optionsJson) {
        return executeWithCancelCheck(() -> {
            String sql = "SELECT task_id, run_id, attempt FROM ssf.spawn_task(?, ?, ?::jsonb, ?::jsonb)";
            try (PreparedStatement cmd = conn.prepareStatement(sql)) {
                cmd.setString(1, queue);
                cmd.setString(2, taskName);
                cmd.setString(3, paramsJson);
                cmd.setString(4, optionsJson);
                try (ResultSet rs = cmd.executeQuery()) {
                    if (rs.next()) {
                        return new SpawnResult(rs.getString(1), rs.getString(2), rs.getInt(3));
                    }
                    throw new RuntimeException("Failed to spawn task");
                }
            }
        });
    }

    @Override
    public void cancelTask(Connection conn, String queue, String taskId) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.cancel_task(?, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(taskId));
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public void emitEvent(Connection conn, String queue, String eventName, String payloadJson) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.emit_event(?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setString(2, eventName);
                cmd.setString(3, payloadJson);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public List<ClaimedTask> claimTasks(Connection conn, String queue, String workerId, int timeout, int count) {
        return executeWithCancelCheck(() -> {
            List<ClaimedTask> tasks = new ArrayList<>();
            String sql = "SELECT run_id, task_id, attempt, task_name, params, retry_strategy, max_attempts, headers, wake_event, event_payload " +
                    "FROM ssf.claim_task(?, ?, ?, ?)";
            try (PreparedStatement cmd = conn.prepareStatement(sql)) {
                cmd.setString(1, queue);
                cmd.setString(2, workerId);
                cmd.setInt(3, timeout);
                cmd.setInt(4, count);

                try (ResultSet rs = cmd.executeQuery()) {
                    while (rs.next()) {
                        int maxAttempts = rs.getInt(7);
                        Integer finalMaxAttempts = rs.wasNull() ? null : maxAttempts;

                        String headersStr = rs.getString(8);
                        ObjectNode headersNode = headersStr != null ? (ObjectNode) mapper.readTree(headersStr) : null;

                        tasks.add(new ClaimedTask(
                                rs.getString(1),      // runId
                                rs.getString(2),      // taskId
                                rs.getString(4),      // taskName
                                rs.getInt(3),         // attempt
                                parseJson(rs, 5),     // params
                                parseJson(rs, 6),     // retryStrategy
                                finalMaxAttempts,     // maxAttempts
                                headersNode,          // headers
                                rs.getString(9),      // wakeEvent
                                parseJson(rs, 10)     // eventPayload
                        ));
                    }
                } catch (JsonMappingException e) {
                    throw new RuntimeException(e);
                } catch (JsonProcessingException e) {
                    throw new RuntimeException(e);
                }
                return tasks;
            }
        });
    }

    @Override
    public void completeRun(Connection conn, String queue, String runId, String resultJson) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.complete_run(?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(runId));
                cmd.setString(3, resultJson);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public void failRun(Connection conn, String queue, String runId, String errorJson) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.fail_run(?, ?, ?::jsonb, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(runId));
                cmd.setString(3, errorJson);
                cmd.setNull(4, Types.TIMESTAMP);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public List<CheckpointRow> getCheckpointStates(Connection conn, String queue, String taskId, String runId) {
        return executeWithCancelCheck(() -> {
            List<CheckpointRow> rows = new ArrayList<>();
            String sql = "SELECT checkpoint_name, state, status, owner_run_id, updated_at " +
                    "FROM ssf.get_task_checkpoint_states(?, ?, ?)";
            try (PreparedStatement cmd = conn.prepareStatement(sql)) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(taskId));
                cmd.setObject(3, UUID.fromString(runId));

                try (ResultSet rs = cmd.executeQuery()) {
                    while (rs.next()) {
                        Timestamp ts = rs.getTimestamp(5);
                        rows.add(new CheckpointRow(
                                rs.getString(1),      // checkpointName
                                parseJson(rs, 2),     // state
                                rs.getString(3),      // status
                                rs.getString(4),      // ownerRunId
                                ts != null ? ts.toInstant() : null // updatedAt
                        ));
                    }
                }
                return rows;
            }
        });
    }

    @Override
    public JsonNode getSingleCheckpoint(Connection conn, String queue, String taskId, String checkpointName) {
        return executeWithCancelCheck(() -> {
            String sql = "SELECT checkpoint_name, state, status, owner_run_id, updated_at " +
                    "FROM ssf.get_task_checkpoint_state(?, ?, ?, ?)";
            try (PreparedStatement cmd = conn.prepareStatement(sql)) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(taskId));
                cmd.setString(3, checkpointName);
                cmd.setInt(4, 0);

                try (ResultSet rs = cmd.executeQuery()) {
                    if (rs.next()) {
                        return parseJson(rs, 2);
                    }
                    return null;
                }
            }
        });
    }

    @Override
    public void persistCheckpoint(Connection conn, String queue, String taskId, String runId, String checkpointName, String stateJson, int timeout) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.set_task_checkpoint_state(?, ?, ?, ?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(taskId));
                cmd.setString(3, checkpointName);
                cmd.setString(4, stateJson);
                cmd.setObject(5, UUID.fromString(runId));
                cmd.setInt(6, timeout);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public void scheduleRun(Connection conn, String queue, String runId, Instant wakeAt) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.schedule_run(?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(runId));
                cmd.setTimestamp(3, Timestamp.from(wakeAt));
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public void heartbeat(Connection conn, String queue, String runId, int seconds) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.extend_claim(?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(runId));
                cmd.setInt(3, seconds);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public EventResult awaitEvent(Connection conn, String queue, String taskId, String runId, String checkpointName, String eventName, Integer timeout) {
        return executeWithCancelCheck(() -> {
            String sql = "SELECT should_suspend, payload FROM ssf.await_event(?, ?, ?, ?, ?, ?)";
            try (PreparedStatement cmd = conn.prepareStatement(sql)) {
                cmd.setString(1, queue);
                cmd.setObject(2, UUID.fromString(taskId));
                cmd.setObject(3, UUID.fromString(runId));
                cmd.setString(4, checkpointName);
                cmd.setString(5, eventName);
                if (timeout != null) cmd.setInt(6, timeout);
                else cmd.setNull(6, Types.INTEGER);

                try (ResultSet rs = cmd.executeQuery()) {
                    if (rs.next()) {
                        return new EventResult(rs.getBoolean(1), parseJson(rs, 2));
                    }
                    throw new RuntimeException("Failed to await event");
                }
            }
        });
    }

    @Override
    public void releaseWorkerClaims(Connection conn, String queue, String workerId) {
        executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("CALL ssf.release_worker_claims(?, ?)")) {
                cmd.setString(1, queue);
                cmd.setString(2, workerId);
                cmd.executeUpdate();
                return null;
            }
        });
    }

    @Override
    public Integer cleanupTasks(Connection conn, String queue, int ttlSeconds, int limit) {
        return executeWithCancelCheck(() -> {
            try (PreparedStatement cmd = conn.prepareStatement("SELECT deleted_tasks FROM ssf.cleanup_tasks(?, ?, ?)")) {
                cmd.setString(1, queue);
                cmd.setInt(2, ttlSeconds);
                cmd.setInt(3, limit);
                try (ResultSet rs = cmd.executeQuery()) {
                    if (rs.next()) return rs.getInt(1);
                    return 0;
                }
            }
        });
    }

    @Override
    public Integer cleanupEvents(Connection conn, String queue, int ttlSeconds, int limit) {
        return cleanupTasks(conn, queue, ttlSeconds, limit);
    }

    private JsonNode parseJson(ResultSet rs, int ordinal) throws SQLException {
        String json = rs.getString(ordinal);
        if (rs.wasNull() || json == null) {
            return null;
        }
        try {
            return mapper.readTree(json);
        } catch (JsonProcessingException e) {
            throw new SQLException("Failed to parse JSON", e);
        }
    }

    @FunctionalInterface
    private interface SqlAction<T> {
        T execute() throws SQLException;
    }

    private <T> T executeWithCancelCheck(SqlAction<T> action) {
        if (Thread.currentThread().isInterrupted()) {
            throw new CancelledTaskException();
        }
        try {
            return action.execute();
        } catch (PSQLException ex) {
            if ("50011".equals(ex.getSQLState())) {
                throw new CancelledTaskException();
            }
            throw new RuntimeException(ex);
        } catch (SQLException ex) {
            throw new RuntimeException(ex);
        }
    }
}

File: sdks/dotnet/src/SqlFlowSdk.Postgres/Database/PostgresFlowDatabase.cs
using Npgsql;
using NpgsqlTypes;
using SqlFlowSdk.Core;
using SqlFlowSdk.Exceptions;
using System.Data;
using System.Data.Common;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SqlFlowSdk.Database;

/// <summary>
/// Encapsulates all raw database interactions. It is used to perform all the necessary operations on the database to 
/// manage queues, tasks, checkpoints, and events in the SqlFlow system using PostgreSQL.
/// </summary>
public class PostgresFlowDatabase : ISqlFlowDatabase
{
    public async Task CreateQueueAsync(DbConnection conn, string queueName, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.create_queue(@p_queue_name, 'unpartitioned')", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queueName);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task DropQueueAsync(DbConnection conn, string queueName, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.drop_queue(@p_queue_name)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queueName);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task<IEnumerable<string>> ListQueuesAsync(DbConnection conn, CancellationToken cancellationToken)
    {
        List<string> results = new();

        using NpgsqlCommand cmd = new("SELECT queue_name FROM ssf.queues ORDER BY queue_name", (NpgsqlConnection)conn);

        using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);

        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            results.Add(reader.GetString(0));
        }
        return results;
    }

    public async Task<SpawnResult> SpawnTaskAsync(DbConnection conn, string queue, string taskName, string paramsJson, string optionsJson, CancellationToken cancellationToken)
    {
        // We cast parameters to jsonb inside the SQL string so PostgreSQL parses the text correctly
        string sql = "SELECT task_id, run_id, attempt FROM ssf.spawn_task(@p_queue_name, @p_task_name, @p_params::jsonb, @p_options::jsonb)";
        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_task_name", NpgsqlDbType.Text, taskName);
        AddParam(cmd, "@p_params", NpgsqlDbType.Jsonb, paramsJson);
        AddParam(cmd, "@p_options", NpgsqlDbType.Jsonb, optionsJson);

        using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);

        if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            return new SpawnResult
            {
                TaskId = reader.GetGuid(0).ToString(),
                RunId = reader.GetGuid(1).ToString(),
                Attempt = reader.GetInt32(2)
            };
        }
        throw new Exception("Failed to spawn task");
    }

    public async Task CancelTaskAsync(DbConnection conn, string queue, string taskId, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.cancel_task(@p_queue_name, @p_task_id)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_task_id", NpgsqlDbType.Uuid, Guid.Parse(taskId));

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task EmitEventAsync(DbConnection conn, string queue, string eventName, string payloadJson, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.emit_event(@p_queue_name, @p_event_name, @p_payload)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_event_name", NpgsqlDbType.Text, eventName);
        AddParam(cmd, "@p_payload", NpgsqlDbType.Jsonb, payloadJson);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task<IEnumerable<ClaimedTask>> ClaimTasksAsync(DbConnection conn, string queue, string workerId, int timeout, int count, CancellationToken cancellationToken)
    {
        List<ClaimedTask> tasks = new();

        string sql = "SELECT run_id, task_id, attempt, task_name, params, retry_strategy, max_attempts, headers, wake_event, event_payload " +
                     "FROM ssf.claim_task(@p_queue_name, @p_worker_id, @p_claim_timeout, @p_qty)";

        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_worker_id", NpgsqlDbType.Text, workerId);
        AddParam(cmd, "@p_claim_timeout", NpgsqlDbType.Integer, timeout);
        AddParam(cmd, "@p_qty", NpgsqlDbType.Integer, count);

        using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);

        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            tasks.Add(new ClaimedTask
            {
                RunId = reader.GetGuid(0).ToString(),
                TaskId = reader.GetGuid(1).ToString(),
                Attempt = reader.GetInt32(2),
                TaskName = reader.GetString(3),
                Params = ParseJson(reader, 4),
                RetryStrategy = ParseJson(reader, 5),
                MaxAttempts = reader.IsDBNull(6) ? null : reader.GetInt32(6),
                Headers = reader.IsDBNull(7) ? null : JsonSerializer.Deserialize<JsonObject>(reader.GetString(7)),
                WakeEvent = reader.IsDBNull(8) ? null : reader.GetString(8),
                EventPayload = ParseJson(reader, 9)
            });
        }

        return tasks;
    }

    public async Task CompleteRunAsync(DbConnection conn, string queue, string runId, string resultJson, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.complete_run(@p_queue_name, @p_run_id, @p_state)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));
        AddParam(cmd, "@p_state", NpgsqlDbType.Jsonb, resultJson);

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task FailRunAsync(DbConnection conn, string queue, string runId, string errorJson, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.fail_run(@p_queue_name, @p_run_id, @p_reason::jsonb, @p_retry_at)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));
        AddParam(cmd, "@p_reason", NpgsqlDbType.Jsonb, errorJson);
        AddParam(cmd, "@p_retry_at", NpgsqlDbType.TimestampTz, DBNull.Value); // Kann optional als UTC DateTime übergeben werden

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task<IEnumerable<CheckpointRow>> GetCheckpointStatesAsync(DbConnection conn, string queue, string taskId, string runId, CancellationToken cancellationToken)
    {
        List<CheckpointRow> rows = new();

        string sql = "SELECT checkpoint_name, state, status, owner_run_id, updated_at " +
                     "FROM ssf.get_task_checkpoint_states(@p_queue_name, @p_task_id, @p_run_id)";
        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_task_id", NpgsqlDbType.Uuid, Guid.Parse(taskId));
        AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));

        using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);

        while (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            rows.Add(new CheckpointRow
            {
                CheckpointName = reader.GetString(0),
                State = ParseJson(reader, 1),
                Status = reader.GetString(2),
                OwnerRunId = reader.IsDBNull(3) ? null : reader.GetGuid(3).ToString(),
                UpdatedAt = reader.GetDateTime(4).ToUniversalTime()
            });
        }

        return rows;
    }

    public async Task<JsonNode?> GetSingleCheckpointAsync(DbConnection conn, string queue, string taskId, string checkpointName, CancellationToken cancellationToken)
    {
        string sql = "SELECT checkpoint_name, state, status, owner_run_id, updated_at FROM ssf.get_task_checkpoint_state(@p_queue_name, @p_task_id, @p_step_name, @p_include_pending)";

        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_task_id", NpgsqlDbType.Uuid, Guid.Parse(taskId));
        AddParam(cmd, "@p_step_name", NpgsqlDbType.Text, checkpointName);
        AddParam(cmd, "@p_include_pending", NpgsqlDbType.Integer, 0);

        using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);

        if (await reader.ReadAsync(cancellationToken).ConfigureAwait(false))
        {
            return ParseJson(reader, 1);
        }

        return null;
    }

    public async Task PersistCheckpointAsync(DbConnection conn, string queue, string taskId, string runId, string checkpointName, string stateJson, int timeout, CancellationToken cancellationToken)
    {
        await ExecuteWithCancelCheckAsync(async (ct) =>
        {
            using NpgsqlCommand cmd = new("CALL ssf.set_task_checkpoint_state(@p_queue_name, @p_task_id, @p_step_name, @p_state, @p_owner_run, @p_extend_claim_by)", (NpgsqlConnection)conn);

            AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
            AddParam(cmd, "@p_task_id", NpgsqlDbType.Uuid, Guid.Parse(taskId));
            AddParam(cmd, "@p_step_name", NpgsqlDbType.Text, checkpointName);
            AddParam(cmd, "@p_state", NpgsqlDbType.Jsonb, stateJson);
            AddParam(cmd, "@p_owner_run", NpgsqlDbType.Uuid, Guid.Parse(runId));
            AddParam(cmd, "@p_extend_claim_by", NpgsqlDbType.Integer, timeout);

            return await cmd.ExecuteNonQueryAsync(ct).ConfigureAwait(false);
        }, cancellationToken);
    }

    public async Task ScheduleRunAsync(DbConnection conn, string queue, string runId, DateTime wakeAt, CancellationToken cancellationToken)
    {
        using NpgsqlCommand cmd = new("CALL ssf.schedule_run(@p_queue_name, @p_run_id, @p_wake_at)", (NpgsqlConnection)conn);

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));
        AddParam(cmd, "@p_wake_at", NpgsqlDbType.TimestampTz, wakeAt.Kind == DateTimeKind.Utc ? wakeAt : wakeAt.ToUniversalTime());

        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
    }

    public async Task HeartbeatAsync(DbConnection conn, string queue, string runId, int seconds, CancellationToken cancellationToken)
    {
        await ExecuteWithCancelCheckAsync(async (ct) =>
        {
            using NpgsqlCommand cmd = new("CALL ssf.extend_claim(@p_queue_name, @p_run_id, @p_extend_by)", (NpgsqlConnection)conn);

            AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
            AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));
            AddParam(cmd, "@p_extend_by", NpgsqlDbType.Integer, seconds);

            return await cmd.ExecuteNonQueryAsync(ct).ConfigureAwait(false);
        }, cancellationToken);
    }

    public async Task<(bool ShouldSuspend, JsonNode? Payload)> AwaitEventAsync(DbConnection conn, string queue, string taskId, string runId, string checkpointName, string eventName, int? timeout, CancellationToken cancellationToken)
    {
        return await ExecuteWithCancelCheckAsync(async (ct) =>
        {
            string sql = "SELECT should_suspend, payload FROM ssf.await_event(@p_queue_name, @p_task_id, @p_run_id, @p_step_name, @p_event_name, @p_timeout)";

            using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

            AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
            AddParam(cmd, "@p_task_id", NpgsqlDbType.Uuid, Guid.Parse(taskId));
            AddParam(cmd, "@p_run_id", NpgsqlDbType.Uuid, Guid.Parse(runId));
            AddParam(cmd, "@p_step_name", NpgsqlDbType.Text, checkpointName);
            AddParam(cmd, "@p_event_name", NpgsqlDbType.Text, eventName);
            AddParam(cmd, "@p_timeout", NpgsqlDbType.Integer, timeout);

            using NpgsqlDataReader reader = await cmd.ExecuteReaderAsync(ct).ConfigureAwait(false);

            if (await reader.ReadAsync(ct).ConfigureAwait(false))
            {
                return (
                    reader.GetBoolean(0),
                    ParseJson(reader, 1)
                );
            }

            throw new Exception("Failed to await event");
        }, cancellationToken);
    }


    public async Task ReleaseWorkerClaimsAsync(DbConnection conn, string queue, string workerId, CancellationToken cancellationToken)
    {
        string sql = "CALL ssf.release_worker_claims(@p_queue_name, @p_worker_id);";

        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        cmd.CommandText = sql;

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_worker_id", NpgsqlDbType.Text, workerId);

        await cmd.ExecuteNonQueryAsync(cancellationToken);
    }


    public async Task<int> CleanupTasksAsync(DbConnection conn, string queue, int ttlSeconds, int limit, CancellationToken cancellationToken)
    {
        string sql = "SELECT deleted_tasks FROM ssf.cleanup_tasks(@p_queue_name, @p_ttl_seconds, @p_limit);";

        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        cmd.CommandText = sql;
        
        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_ttl_seconds", NpgsqlDbType.Integer, ttlSeconds);
        AddParam(cmd, "@p_limit", NpgsqlDbType.Integer, limit);

        object? result = await cmd.ExecuteScalarAsync(cancellationToken);

        return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
    }

    public async Task<int> CleanupEventsAsync(DbConnection conn, string queue, int ttlSeconds, int limit, CancellationToken cancellationToken)
    {
        string sql = "SELECT deleted_tasks FROM ssf.cleanup_tasks(@p_queue_name, @p_ttl_seconds, @p_limit);";

        using NpgsqlCommand cmd = new(sql, (NpgsqlConnection)conn);

        cmd.CommandText = sql;

        AddParam(cmd, "@p_queue_name", NpgsqlDbType.Text, queue);
        AddParam(cmd, "@p_ttl_seconds", NpgsqlDbType.Integer, ttlSeconds);
        AddParam(cmd, "@p_limit", NpgsqlDbType.Integer, limit);

        object? result = await cmd.ExecuteScalarAsync(cancellationToken);

        return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
    }

    private static JsonNode? ParseJson(NpgsqlDataReader reader, int ordinal)
    {
        if (reader.IsDBNull(ordinal))
        {
            return null;
        }

        return JsonSerializer.Deserialize<JsonNode>(reader.GetString(ordinal));
    }

    private async Task<T> ExecuteWithCancelCheckAsync<T>(Func<CancellationToken, Task<T>> action, CancellationToken ct)
    {
        try
        {
            return await action(ct).ConfigureAwait(false);
        }
        catch (PostgresException ex) when (ex.SqlState == "50011")
        {
            throw new CancelledTaskException();
        }
    }

    private void AddParam(NpgsqlCommand cmd, string name, NpgsqlDbType dbType, object? value)
    {
        cmd.Parameters.AddWithValue(name, dbType, value ?? DBNull.Value);
    }
}